daemon/containerd/image_exporter.go

```diff
@@ -342,3 +378,61 @@ func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, platf
 
 	return nil
 }
+
+// verifyImagesProvidePlatform checks if the requested platform is loaded.
+// If the requested platform is not loaded, it returns an error.
+func (i *ImageService) verifyImagesProvidePlatform(ctx context.Context, imgs []containerdimages.Image, platform ocispec.Platform, pm platforms.Matcher) error {
+	var missingImages []string
+	for _, img := range imgs {
+		hasRequestedPlatform := false
+		err := i.walkImageManifests(ctx, img, func(platformImg *ImageManifest) error {
+			imgPlat, err := platformImg.ImagePlatform(ctx)
+			if err != nil {
+				if cerrdefs.IsNotFound(err) {
+					return nil
+				}
+				return errors.Wrapf(err, "failed to determine image platform")
+			}
+
+			if !pm.Match(imgPlat) {
+				return nil
+			}
+			available, err := platformImg.CheckContentAvailable(ctx)
+			if err != nil {
+				return errors.Wrapf(err,
+					"failed to determine image content availability for platform %s",
+					platforms.FormatAll(platform),
+				)
+			}
+
+			if available {
+				hasRequestedPlatform = true
+				return nil
+			}
+			return nil
+		})
+		if err != nil {
+			return errdefs.System(err)
+		}
+		if !hasRequestedPlatform {
+			missingImages = append(missingImages, imageFamiliarName(img))
+		}
+	}
+
+	allImageNames := strings.Join(missingImages, ", ")
+	platformStr := platforms.FormatAll(platform)
+
+	msg := ""
+	switch len(missingImages) {
+	case 0:
+		// Success - All images provide the requested platform.
+		return nil
+	case 1:
+		msg = "image %s was loaded, but doesn't provide the requested platform (%s)"
```

Comment:
I should probably check the actual code, but was the image actually _loaded_ in this case? Will it be garbage-collected immediately after if it doesn't match what we want?

(in that case, considering if `was loaded` would potentially only add confusion, and if a shorter error would work;


```
image %s doesn't provide the requested platform (%s)
```


We could mention "contained in the archive", but not sure if that works;

```
image %s contained in the archive doesn't provide the requested platform (%s)
```

Reply:
This sounds like a possible tracking ticket in the making 😂 ❤️

Reply:
So I gave this a try with this PR, and while the content is loaded, I _THINK_ it discards the unused ones after? Here's what I did.

I had an alpine image with 2 variants, and exported that one;


```console
$ docker image ls --tree
IMAGE                   ID             DISK USAGE   CONTENT SIZE   USED
alpine:latest           beefdbd8a1da       24.2MB         7.46MB
├─ linux/riscv64        80cde017a105       10.6MB         3.37MB
├─ linux/arm64/v8       9cee2b382fe2       13.6MB         4.09MB
├─ linux/amd64          33735bd63cf8           0B             0B
├─ linux/arm/v6         50f635c8b04d           0B             0B
├─ linux/arm/v7         f2f82d424957           0B             0B
├─ linux/386            b3e87f642f5c           0B             0B
├─ linux/ppc64le        c7a6800e3dc5           0B             0B
└─ linux/s390x          2b5b26e09ca2           0B             0B

$ docker image save -o alpine2.tar alpine:latest
```

I removed the image (daemon now had no images);

```console
$ docker image rm alpine:latest
Untagged: alpine:latest
Deleted: sha256:beefdbd8a1da6d2915566fde36db9db0b524eb737fc57cd1367effd16dc0d06d

$ docker system prune -af
Total reclaimed space: 0B
```

Then loaded the image;

```console
$ docker image load --platform=linux/arm64/v8 --platform=linux/riscv64 -i alpine2.tar
Loaded image: alpine:latest
```

Logs produced as part of this show that content is being removed;

```console
INFO[2024-10-23T07:40:15.747273425Z] API listen on /var/run/docker.sock
DEBU[2024-10-23T07:59:00.084397751Z] handling HEAD request                         method=HEAD module=api request-url=/_ping spanID=7b34e8774792cc0f status=200 traceID=92831a6f26302569d4ea318b0f4c65c4 vars="map[]"
DEBU[2024-10-23T07:59:00.244992084Z] handling POST request                         method=POST module=api request-url="/v1.48/images/load?platform=%7B%22architecture%22%3A%22arm64%22%2C%22os%22%3A%22linux%22%2C%22variant%22%3A%22v8%22%7D&quiet=0" spanID=22be75fe2aca9ff6 status=200 traceID=8e7b378cd832282342c096931cd42c4e vars="map[version:1.48]"
DEBU[2024-10-23T07:59:00.090846709Z] (*service).Write started                      ref=tar-blobs/sha256/2f32f6b11fa159940aadedb7a73a2834f1314ad99e989e38e12dff7ba6575d03 total=1473
DEBU[2024-10-23T07:59:00.093687834Z] (*service).Write started                      ref=tar-blobs/sha256/80cde017a10529a18a7274f70c687bb07c4969980ddfb35a1b921fda3a020e5b total=528
DEBU[2024-10-23T07:59:00.096256418Z] (*service).Write started                      ref=tar-blobs/sha256/8c4a05189a5fd2cf629c25ab8d0831be7156d74b336f129a412933ee78af018c total=3371452
DEBU[2024-10-23T07:59:00.131533543Z] (*service).Write started                      ref=tar-blobs/sha256/9cee2b382fe2412cd77d5d437d15a93da8de373813621f2e4d406e3df0cf0e7c total=528
DEBU[2024-10-23T07:59:00.135708668Z] (*service).Write started                      ref=tar-blobs/sha256/beefdbd8a1da6d2915566fde36db9db0b524eb737fc57cd1367effd16dc0d06d total=1853
DEBU[2024-10-23T07:59:00.139100543Z] (*service).Write started                      ref=tar-blobs/sha256/c157a85ed455142fd79bff5dce951fd5f5b0d0c6e45e6f54cfd0c4e2bdec587b total=1486
DEBU[2024-10-23T07:59:00.141611376Z] (*service).Write started                      ref=tar-blobs/sha256/cf04c63912e16506c4413937c7f4579018e4bb25c272d989789cfba77b12f951 total=4087646
DEBU[2024-10-23T07:59:00.167765126Z] (*service).Write started                      ref=tar-index.json total=441
DEBU[2024-10-23T07:59:00.174570584Z] create image                                  name="docker.io/library/alpine:latest" target="sha256:beefdbd8a1da6d2915566fde36db9db0b524eb737fc57cd1367effd16dc0d06d"
DEBU[2024-10-23T07:59:00.181258376Z] stat snapshot                                 key="sha256:16113d51b7181f20135f51e8ffbaead20a7671cd783017601f198748ce8a8ebf"
DEBU[2024-10-23T07:59:00.182770793Z] stat snapshot                                 key="sha256:16113d51b7181f20135f51e8ffbaead20a7671cd783017601f198748ce8a8ebf"
DEBU[2024-10-23T07:59:00.183003918Z] prepare snapshot                              key="extract-182858084-qVQF sha256:16113d51b7181f20135f51e8ffbaead20a7671cd783017601f198748ce8a8ebf" parent=
DEBU[2024-10-23T07:59:00.185013376Z] using pigz for decompression
DEBU[2024-10-23T07:59:00.234568126Z] diff applied                                  d=49.6165ms digest="sha256:cf04c63912e16506c4413937c7f4579018e4bb25c272d989789cfba77b12f951" media=application/vnd.docker.image.rootfs.diff.tar.gzip size=4087646
DEBU[2024-10-23T07:59:00.235144709Z] commit snapshot                               key="extract-182858084-qVQF sha256:16113d51b7181f20135f51e8ffbaead20a7671cd783017601f198748ce8a8ebf" name="sha256:16113d51b7181f20135f51e8ffbaead20a7671cd783017601f198748ce8a8ebf"
DEBU[2024-10-23T07:59:00.238797918Z] unpack                                        alreadyUnpacked=false error="<nil>" image="alpine:latest" manifest="sha256:9cee2b382fe2412cd77d5d437d15a93da8de373813621f2e4d406e3df0cf0e7c" spanID=22be75fe2aca9ff6 traceID=8e7b378cd832282342c096931cd42c4e
DEBU[2024-10-23T07:59:00.367122126Z] remove content                                key="sha256:2f32f6b11fa159940aadedb7a73a2834f1314ad99e989e38e12dff7ba6575d03"
DEBU[2024-10-23T07:59:00.367161501Z] remove content                                key="sha256:355eee6af939abf5ba465c9be69c3b725f8d3f19516ca9644cf2a4fb112fd83b"
DEBU[2024-10-23T07:59:00.367171876Z] remove content                                key="sha256:80cde017a10529a18a7274f70c687bb07c4969980ddfb35a1b921fda3a020e5b"
DEBU[2024-10-23T07:59:00.367177876Z] remove content                                key="sha256:8c4a05189a5fd2cf629c25ab8d0831be7156d74b336f129a412933ee78af018c"
DEBU[2024-10-23T07:59:00.367410501Z] schedule content cleanup
DEBU[2024-10-23T07:59:00.367574335Z] removed content                               digest="sha256:2f32f6b11fa159940aadedb7a73a2834f1314ad99e989e38e12dff7ba6575d03"
DEBU[2024-10-23T07:59:00.367631835Z] removed content                               digest="sha256:355eee6af939abf5ba465c9be69c3b725f8d3f19516ca9644cf2a4fb112fd83b"
DEBU[2024-10-23T07:59:00.367656835Z] removed content                               digest="sha256:80cde017a10529a18a7274f70c687bb07c4969980ddfb35a1b921fda3a020e5b"
DEBU[2024-10-23T07:59:00.367862751Z] removed content                               digest="sha256:8c4a05189a5fd2cf629c25ab8d0831be7156d74b336f129a412933ee78af018c"
```

And `docker image ls --tree` only shows one image taking up space:

```console
$ docker image ls --tree
IMAGE                   ID             DISK USAGE   CONTENT SIZE   USED
alpine:latest           beefdbd8a1da       13.6MB         4.09MB
├─ linux/amd64          33735bd63cf8           0B             0B
├─ linux/arm/v6         50f635c8b04d           0B             0B
├─ linux/arm/v7         f2f82d424957           0B             0B
├─ linux/arm64/v8       9cee2b382fe2       13.6MB         4.09MB
├─ linux/386            b3e87f642f5c           0B             0B
├─ linux/ppc64le        c7a6800e3dc5           0B             0B
├─ linux/riscv64        80cde017a105           0B             0B
└─ linux/s390x          2b5b26e09ca2           0B             0B
```

So I _think_ we're good, but yeah, would be nice if we could skip the actual `load` -> `remove` part.

Reply:
In this case it will work correctly - I was talking about a case where the load fails because the archive didn't contain the requested platform. For example in the case above (without this PR):

```bash
$ docker load --platform linux/s390x -i alpine2.tar 
Loaded image: alpine:latest

$ docker images --tree
WARNING: This is an experimental feature. The output may change and shouldn't be depended on.

IMAGE           ID             DISK USAGE   CONTENT SIZE   IN USE
alpine:latest   beefdbd8a1da           0B             0B        
```

The image is empty, because only the index got loaded and the image was already created by `client.Import`.

Reply:
Ah! Alright, so yes, with this PR that works correctly;

```bash
docker load --platform linux/s390x -i alpine2.tar
requested platform (linux/s390x) found, but some content is missing: NotFound: content digest sha256:2b5b26e09ca2856f50ac88312348d26c1ac4b8af1df9f580e5cf465fd76e3d4d: not found

docker image ls --tree
IMAGE   ID             DISK USAGE   CONTENT SIZE   IN USE

docker load -i alpine2.tar
Loaded image: alpine:latest

docker image ls --tree
IMAGE                   ID             DISK USAGE   CONTENT SIZE   IN USE
alpine:latest           beefdbd8a1da       16.9MB         7.46MB
├─ linux/amd64          33735bd63cf8           0B             0B
├─ linux/arm/v6         50f635c8b04d           0B             0B
├─ linux/arm/v7         f2f82d424957           0B             0B
├─ linux/arm64/v8       9cee2b382fe2       13.6MB         4.09MB
├─ linux/386            b3e87f642f5c           0B             0B
├─ linux/ppc64le        c7a6800e3dc5           0B             0B
├─ linux/riscv64        80cde017a105       3.37MB         3.37MB
└─ linux/s390x          2b5b26e09ca2           0B             0B
```

Reply:
Tracking ticket;

- https://github.com/moby/moby/issues/48762